From 81031292fb21e76e1a5aed5b29c1e534857b10e8 Mon Sep 17 00:00:00 2001
From: Antheas Kapenekakis <lkml@antheas.dev>
Date: Fri, 24 Oct 2025 17:56:17 +0200
Subject: [PATCH 045/112] platform/x86: asus-wmi: add early backlight init
 quirk

The Asus Zenbook Duo's keyboard is detacheable and when it is detached
there is no led device initialized. This prevents userspace from probing
for backlight brightness support on boot. Add a quirk to always
initialize the backlight device early so that it is always available.

Signed-off-by: Antheas Kapenekakis <lkml@antheas.dev>
---
 drivers/platform/x86/asus-nb-wmi.c | 1 +
 drivers/platform/x86/asus-wmi.c    | 4 +++-
 drivers/platform/x86/asus-wmi.h    | 1 +
 3 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/platform/x86/asus-nb-wmi.c b/drivers/platform/x86/asus-nb-wmi.c
index a38a65f5c..da0e805ef 100644
--- a/drivers/platform/x86/asus-nb-wmi.c
+++ b/drivers/platform/x86/asus-nb-wmi.c
@@ -148,6 +148,7 @@ static struct quirk_entry quirk_asus_ignore_fan = {
 
 static struct quirk_entry quirk_asus_zenbook_duo_kbd = {
 	.key_wlan_event = ASUS_WMI_KEY_IGNORE,
+	.init_backlight_early = true,
 };
 
 static struct quirk_entry quirk_asus_z13 = {
diff --git a/drivers/platform/x86/asus-wmi.c b/drivers/platform/x86/asus-wmi.c
index e79289d61..d2f7f48d8 100644
--- a/drivers/platform/x86/asus-wmi.c
+++ b/drivers/platform/x86/asus-wmi.c
@@ -2052,7 +2052,9 @@ static int asus_wmi_led_init(struct asus_wmi *asus)
 
 	scoped_guard(spinlock_irqsave, &asus_ref.lock) {
 		asus_ref.asus = asus;
-		if (asus->kbd_led_avail || !list_empty(&asus_ref.listeners))
+		if (asus->kbd_led_avail ||
+		    asus->driver->quirks->init_backlight_early ||
+		    !list_empty(&asus_ref.listeners))
 			queue_work(asus->led_workqueue, &asus->kbd_led_work);
 	}
 
diff --git a/drivers/platform/x86/asus-wmi.h b/drivers/platform/x86/asus-wmi.h
index 5cd4392b9..73710ae9a 100644
--- a/drivers/platform/x86/asus-wmi.h
+++ b/drivers/platform/x86/asus-wmi.h
@@ -41,6 +41,7 @@ struct quirk_entry {
 	bool wmi_force_als_set;
 	bool wmi_ignore_fan;
 	bool filter_i8042_e1_extended_codes;
+	bool init_backlight_early;
 	int key_wlan_event;
 	enum asus_wmi_tablet_switch_mode tablet_switch_mode;
 	int wapf;
-- 
2.52.0

